#!/bin/bash

source activate qiime2-2018.8

classifier_ref=$DB_REF/gg-13-8-99-515-806-nb-classifier.qza

reads=EMPPairedEndSequences

while getopts ":r:ph" opt;
do
    case ${opt} in	
      h) echo "qiime2 pipeline for analysis single-end data generated by Illumina Miseq"
	 echo "Usage:"
	 echo "    qiime2.sh -h               Display this help message."
	 echo "    qimmer2.sh -r  <reads directory>"  
	 exit 0
      ;;
      :) "Invalid option: $OPTARG requires an argument" 1>&2
      ;;
      p) reads=EMPSingleEndSequences
      ;;
      r) raw_reads_dir=${OPTARG%/}
	 
	 if [ ! -d $raw_reads_dir ]
	 then
	     echo "Failed to locate the reads directory: $raw_reads_dir "
	 fi
      ;;
      \?) echo "Usage: cmd [-h] [-r] [s]"
      ;;
   esac
done
shift $((OPTIND -1))



if [ -z $raw_reads_dir ]
   
then
    echo "Invalid option: fastq reads directory [-r] is required" 1>&2
    echo "Usage:"
    echo "Usage:"
    echo "    qiime2.sh -h   Display this help message."
    echo "    qimmer2.sh -r  <reads directory>"  
    exit 1
fi


if [ ! -f $classifier_ref ];
then
    echo "File not found!"
    exit  0
fi


echo $raw_reads_dir
echo $reads


##importing data
qiime tools import \
      --type $reads \
      --input-path $raw_reads_dir \
      --output-path emp-single-end-sequences.qza



##Demultiplexing sequences
qiime demux emp-single \
      --i-seqs emp-single-end-sequences.qza \
      --m-barcodes-file sample-metadata.tsv \
      --m-barcodes-column BarcodeSequence \
      --o-per-sample-sequences demux.qza



##generate a summary of the demultiplexing results
qiime demux summarize \
      --i-data demux.qza \
      --o-visualization demux.qzv




## Dada2 

qiime dada2 denoise-single \
      --i-demultiplexed-seqs demux.qza \
      --p-n-threads 0\
      --p-trim-left 0 \
      --p-trunc-len 120 \
      --o-representative-sequences rep-seqs-dada2.qza \
      --o-table table-dada2.qza \
      --o-denoising-stats dada2-stats.qza\
      --verbose



qiime metadata tabulate \
      --m-input-file dada2-stats.qza \
      --o-visualization dada2-stats.qzv


##Deblur
qiime quality-filter q-score \
      --i-demux demux.qza \
      --o-filtered-sequences demux-filtered.qza \
      --o-filter-stats demux-filter-stats.qza\
      --verbose




qiime deblur denoise-16S \
      --i-demultiplexed-seqs demux-filtered.qza \
      --p-trim-length 120 \
      --o-representative-sequences rep-seqs-deblur.qza \
      --o-table table-deblur.qza \
      --p-sample-stats \
      --o-stats deblur-stats.qza\
      --p-jobs-to-start 2\
      --verbose





qiime metadata tabulate \
      --m-input-file demux-filter-stats.qza \
      --o-visualization demux-filter-stats.qzv




qiime deblur visualize-stats \
      --i-deblur-stats deblur-stats.qza \
        --o-visualization deblur-stats.qzv


# cp rep-seqs-dada2.qza rep-seqs.qza
# cp table-dada2.qza table.qza


cp rep-seqs-deblur.qza rep-seqs.qza
cp table-deblur.qza table.qza


#FeatureTable and FeatureData summaries

qiime feature-table summarize \
      --i-table table.qza \
      --o-visualization table.qzv \
      --m-sample-metadata-file sample-metadata.tsv




qiime feature-table tabulate-seqs \
      --i-data rep-seqs.qza \
        --o-visualization rep-seqs.qzv



qiime phylogeny align-to-tree-mafft-fasttree\
      --p-n-threads 0\
      --i-sequences rep-seqs.qza \
      --o-alignment aligned-rep-seqs.qza \
      --o-masked-alignment masked-aligned-rep-seqs.qza \
      --o-tree unrooted-tree.qza \
      --o-rooted-tree rooted-tree.qza\
      --verbose




qiime diversity core-metrics-phylogenetic \
      --i-phylogeny rooted-tree.qza \
      --i-table table.qza \
      --p-sampling-depth 1109 \
      --m-metadata-file sample-metadata.tsv \
      --output-dir core-metrics-results\
      --verbose



qiime diversity alpha-group-significance \
      --i-alpha-diversity core-metrics-results/faith_pd_vector.qza \
      --m-metadata-file sample-metadata.tsv \
      --o-visualization core-metrics-results/faith-pd-group-significance.qzv\
      --verbose




qiime diversity alpha-group-significance \
      --i-alpha-diversity core-metrics-results/evenness_vector.qza \
      --m-metadata-file sample-metadata.tsv \
      --o-visualization core-metrics-results/evenness-group-significance.qzv




qiime diversity beta-group-significance \
      --i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza \
      --m-metadata-file sample-metadata.tsv \
      --m-metadata-column BodySite \
      --o-visualization core-metrics-results/unweighted-unifrac-body-site-significance.qzv \
      --p-pairwise




qiime diversity beta-group-significance \
      --i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza \
      --m-metadata-file sample-metadata.tsv \
      --m-metadata-column Subject \
      --o-visualization core-metrics-results/unweighted-unifrac-subject-group-significance.qzv \
      --p-pairwise




qiime emperor plot \
      --i-pcoa core-metrics-results/unweighted_unifrac_pcoa_results.qza \
      --m-metadata-file sample-metadata.tsv \
      --p-custom-axes DaysSinceExperimentStart \
      --o-visualization core-metrics-results/unweighted-unifrac-emperor-DaysSinceExperimentStart.qzv




qiime emperor plot \
      --i-pcoa core-metrics-results/bray_curtis_pcoa_results.qza \
      --m-metadata-file sample-metadata.tsv \
      --p-custom-axes DaysSinceExperimentStart \
      --o-visualization core-metrics-results/bray-curtis-emperor-DaysSinceExperimentStart.qzv




qiime diversity alpha-rarefaction \
      --i-table table.qza \
      --i-phylogeny rooted-tree.qza \
      --p-max-depth 4000 \
      --m-metadata-file sample-metadata.tsv \
      --o-visualization alpha-rarefaction.qzv




qiime feature-classifier classify-sklearn \
      --i-classifier $classifier_ref \
      --i-reads rep-seqs.qza \
      --o-classification taxonomy.qza




qiime metadata tabulate \
      --m-input-file taxonomy.qza \
      --o-visualization taxonomy.qzv




qiime taxa barplot \
      --i-table table.qza \
      --i-taxonomy taxonomy.qza \
      --m-metadata-file sample-metadata.tsv \
      --o-visualization taxa-bar-plots.qzv




qiime feature-table filter-samples \
      --i-table table.qza \
      --m-metadata-file sample-metadata.tsv \
      --p-where "BodySite='gut'" \
      --o-filtered-table gut-table.qza




qiime composition add-pseudocount \
      --i-table gut-table.qza \
      --o-composition-table comp-gut-table.qza




qiime composition ancom \
      --i-table comp-gut-table.qza \
      --m-metadata-file sample-metadata.tsv \
      --m-metadata-column Subject \
      --o-visualization ancom-Subject.qzv




qiime taxa collapse \
      --i-table gut-table.qza \
      --i-taxonomy taxonomy.qza \
      --p-level 6 \
      --o-collapsed-table gut-table-l6.qza




qiime composition add-pseudocount \
      --i-table gut-table-l6.qza \
      --o-composition-table comp-gut-table-l6.qza




qiime composition ancom \
      --i-table comp-gut-table-l6.qza \
      --m-metadata-file sample-metadata.tsv \
      --m-metadata-column Subject \
      --o-visualization l6-ancom-Subject.qzv
